src/ui/public/visualize/visualization.js

```diff
@@ -114,16 +121,25 @@ uiModules
           renderSubscription.unsubscribe();
         });
 
-        if (!$scope.vis.visualizeScope) {
-          $scope.$watchGroup(['visData', 'vis.params'], () => {
+        // Listen on changes to trigger a render if listenOnChange is not false
+        // i.e. true or has not been used (undefined in that case)
+        if ($scope.listenOnChange !== false) {
+          const onChangeListener = () => {
             $scope.$emit('render');
-          });
+          };
 
-          resizeChecker.on('resize',  () => {
-            $scope.$emit('render');
+          $scope.$watchGroup(['visData', 'vis.params'], onChangeListener);
+
+          $scope.uiState.on('change', onChangeListener);
+          $scope.$on('$destroy', () => {
+            $scope.uiState.off('change', onChangeListener);
           });
         }
 
+        resizeChecker.on('resize', () => {
```

Comment:
We now do the resize checking of the chart element always (also when running inside `visualize`). This wasn't done earlier, since `visualize` would look on its size. But that wouldn't help detecting changes to the actual chart div, as described above, why we changed the element on which the resize checker is running.